### PR TITLE
Cache sounds

### LINKED ISSUE #89: Need to cache sounds
We cache boards and images but not sounds currently

### PR DESCRIPTION
**Pull request by blaise17**
Closes #89 

Test file with path, data and url image: https://dl.dropboxusercontent.com/s/8ihech9oq8pz00u/test4.obz?dl=1

### CODE REVIEW

File: src/app/obzboard-set.ts [hunk 1]
```diff
@@ -61,20 +61,22 @@ export class OBZBoardSet implements ImageResolver, SoundResolver {
...
-    progress.progress(ProgressService.message('Pre-caching images'));
-    return this.blobifyImages(httpClient).pipe(
-      map(imageResult => this)
+    return forkJoin(this.blobifyImages(httpClient, progress), this.blobifySounds(httpClient, progress)).pipe(
+      map(result => this)
     );
+  }
 
-    // TODO: sounds
+  private progressObservable(message: string, progress: ProgressService): Observable<boolean> {
```
Comment (psi77) [inline src/app/obzboard-set.ts:69]: Should this be static?

File: src/app/obzboard-set.ts [hunk 2]
```diff
@@ -106,6 +108,53 @@ export class OBZBoardSet implements ImageResolver, SoundResolver {
     return observables.length > 0 ? forkJoin(...observables) : of(true);
   }
 
+  private blobifySounds(httpClient: HttpClient, progress: ProgressService): Observable<boolean> {
+
+    const observables = [];
+
+    this.boards.forEach((board: OBFBoard) => {
+      observables.push(this.progressObservable('Pre-caching sounds.', progress));
```
Comment (psi77) [inline src/app/obzboard-set.ts:116]: Should this be outside of the forEach loop?

Comment (psi77) [reply src/app/obzboard-set.ts:116]: or do we want to add the board id to the message and keep it inside the loop?

File: src/app/obzboard-set.ts [hunk 3]
```diff
@@ -106,6 +108,53 @@ export class OBZBoardSet implements ImageResolver, SoundResolver {
...
+            const key = `data:${sound.id}`;
+            // TODO: we could use the content type from here that we're stripping off
+            const data = sound.data.substr(sound.data.indexOf(',') + 1);
+            this.setSound(key, data);
+            // set path to key so render code will look up blob in sounds
+            sound.path = key;
+            // remove data as we don't need to cache it too
+            sound.data = null;
+          } else if (sound.url) {
+            // load url into blob
```
Comment (psi77) [inline src/app/obzboard-set.ts:131]: Technically this is loading to base64 string rather than a blob

File: src/app/obzboard-set.ts [hunk 4]
```diff
@@ -106,6 +108,53 @@ export class OBZBoardSet implements ImageResolver, SoundResolver {
...
+            console.log(`Loading sound ${sound.url}`);
+            observables.push(httpClient.get(sound.url, { responseType: 'blob' }).pipe(
+              mergeMap((blob: Blob): Observable<Event> => {
+                const reader = new FileReader();
```
Comment (psi77) [inline src/app/obzboard-set.ts:135]: Probably worth raising a follow on ticket for the error handling
(maybe something like https://gist.github.com/iansinnott/3d0ba1e9edc3e6967bc51da7020926b0)

**Review by psi77 — APPROVED** [review]
Looks great!

File: src/app/obzboard-set.ts (hunk 2 above)
Comment (blaise17) [reply src/app/obzboard-set.ts]: I think it should be outside... unless we get feedback that they're too slow and need more granularity?

File: src/app/obzboard-set.ts (hunk 4 above)
Comment (blaise17) [reply src/app/obzboard-set.ts:135]: Done

Comment (blaise17) [reply src/app/obzboard-set.ts:135]: Issue #154

**Review by psi77 — APPROVED** [review]
Still looks great!

**Review by psi77 — COMMENTED** [review]
Test comment

**Pull request merged**